internal/queryexecutor/executor.go

```diff
@@ -0,0 +1,160 @@
+	if err != nil {
+		return nil, err
+	}
+
+	if marshledIndexDef == nil {
+		return nil, errors.New("no index has been defined on the database " + dbName)
+	}
+
+	indexDef := map[string]types.Type{}
+	if err := json.Unmarshal(marshledIndexDef, &indexDef); err != nil {
+		return nil, err
+	}
+
+	queryConditions := make(attributesConditions)
+	for attr, c := range conditions {
+		if _, ok := indexDef[attr]; !ok {
+			return nil, errors.New("attribute [" + attr + "] given in the query condition is not indexed")
+		}
+
+		cond, ok := c.(map[string]interface{})
+		if !ok {
+			return nil, errors.New("query syntax error near the attribute [" + attr + "]")
+		}
+
+		attrType := indexDef[attr]
+		conds := &attrConditions{
+			valueType:  attrType,
+			conditions: make(map[string]interface{}),
+		}
+
+		for opt, v := range cond {
+			if !isValidLogicalOperator(opt) {
+				return nil, errors.New("invalid logical operator [" + opt + "] provided for the attribute [" + attr + "]")
+			}
+
+			if err := validateType(v, attrType); err != nil {
+				return nil, errors.WithMessage(err, "attribute ["+attr+"] is indexed but the value type provided in the query does not match the actual indexed type")
+			}
+
+			if attrType == types.Type_NUMBER {
+				v, err = v.(json.Number).Int64()
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			conds.conditions[opt] = v
+		}
+
+		if err := validateAttrConditions(conds.conditions); err != nil {
+			return nil, errors.WithMessage(err, "query syntax error near attribute ["+attr+"]")
+		}
+		queryConditions[attr] = conds
+	}
+
+	return queryConditions, nil
+}
+
+func isValidLogicalOperator(opt string) bool {
+	switch opt {
+	case constants.Equal,
+		constants.NotEqual,
+		constants.GreaterThan,
+		constants.LesserThan,
+		constants.GreaterThanOrEqual,
+		constants.LesserThanOrEqual:
+		return true
+	default:
+		return false
+	}
+}
+
+func validateType(v interface{}, t types.Type) error {
+	kind := reflect.TypeOf(v).Kind()
+	switch kind {
+	case reflect.String:
+		isNumber := reflect.TypeOf(v).Name() == "Number"
+
+		if t == types.Type_STRING && !isNumber {
+			return nil
+		} else if t == types.Type_NUMBER && isNumber {
+			return nil
+		} else {
+			providedType := kind.String()
+			if isNumber {
+				providedType = "number"
+			}
+			return errors.New("the actual type [" + strings.ToLower(t.String()) + "]" +
+				" does not match the provided type [" + providedType + "]")
+		}
+
+	case reflect.Bool:
+		if t == types.Type_BOOLEAN {
+			return nil
+		}
+		return errors.New("the actual type [" + strings.ToLower(t.String()) + "]" +
+			" does not match the provided type [" + kind.String() + "]")
+	default:
+		return errors.New("the actual type [" + strings.ToLower(t.String()) + "]" +
+			" does not match the provided type [" + kind.String() + "]")
+	}
+}
+
+func validateAttrConditions(conds map[string]interface{}) error {
+	if _, ok := conds[constants.Equal]; ok {
```

Comment:
what about `NotEqual`?

Reply:
Good catch. Currently, notEqual condition is supported yet. It requires a bit more complex range query plan. Added error for not equal to condition.